lib/src/conflicts.rs

```diff
@@ -454,25 +483,46 @@ pub fn update_conflict_from_content(
         // with conflict markers. Update the Conflict object with the new
         // FileIds.
         for (i, buf) in removed_content.iter().enumerate() {
-            let file_id = store.write_file(path, &mut buf.as_slice())?;
-            if let TreeValue::File { id, executable: _ } = &mut conflict.removes[i].value {
-                *id = file_id;
-            } else {
-                // TODO: This can actually happen. We should check earlier
-                // that the we only attempt to parse the conflicts if it's a
-                // file-only conflict.
-                panic!("Found conflict markers in merge of non-files");
+            match &conflict.removes()[i] {
+                Some(TreeValue::File { id: _, executable }) => {
+                    let file_id = store.write_file(path, &mut buf.as_slice())?;
+                    let new_value = TreeValue::File {
+                        id: file_id,
+                        executable: *executable,
+                    };
+                    conflict.set_remove(i, Some(new_value));
+                }
+                None if buf.is_empty() => {
+                    // The missing side of a conflict is still represented by
+                    // the empty string we materialized it as => nothing to do
+                }
+                _ => {
+                    // The user edited a non-file side. We consider the conflict resolved.
+                    return Ok(None);
+                }
             }
         }
         for (i, buf) in added_content.iter().enumerate() {
-            let file_id = store.write_file(path, &mut buf.as_slice())?;
-            if let TreeValue::File { id, executable: _ } = &mut conflict.adds[i].value {
-                *id = file_id;
-            } else {
-                panic!("Found conflict markers in merge of non-files");
+            match &conflict.adds()[i] {
+                Some(TreeValue::File { id: _, executable }) => {
+                    let file_id = store.write_file(path, &mut buf.as_slice())?;
+                    let new_value = TreeValue::File {
+                        id: file_id,
+                        executable: *executable,
+                    };
+                    conflict.set_add(i, Some(new_value));
+                }
+                None if buf.is_empty() => {
+                    // The missing side of a conflict is still represented by
+                    // the empty string we materialized it as => nothing to do
+                }
+                _ => {
+                    // The user edited a non-file side. We consider the conflict resolved.
+                    return Ok(None);
```

Comment:
Same as above.

Also, very optional: perhaps the inside of the for loop should be wrapped in a closure to reduce duplication. I'm not sure whether ownership rules would make this a pain.